chapters/slang.adoc

```diff
@@ -0,0 +1,744 @@
+
+=== Interfaces and Polymorphism
+
+Slang introduces interfaces, enabling polymorphic behavior in shaders:
+
+[source,slang]
+----
+// Define an interface
+interface IMaterial {
+    float3 evaluateBRDF(float3 viewDir, float3 lightDir, float3 normal);
+}
+
+// Implement the interface
+struct LambertianMaterial : IMaterial {
+    float3 albedo;
+
+    float3 evaluateBRDF(float3 viewDir, float3 lightDir, float3 normal) {
+        return albedo / 3.14159;
+    }
+}
+
+// Use polymorphically
+void shadeSurface(IMaterial material, ...) {
+    // Use the material interface without knowing the concrete type
+}
+----
+
+This feature is not available in standard HLSL.
+
+=== Modules and Namespaces
+
+Slang provides a module system for better code organization:
+
+[source,slang]
+----
+// In file: lighting.slang
+module Lighting;
+
+// Export functions to be used by other modules
+export float3 calculateDirectLighting(...) { ... }
+
+// In file: main.slang
+import Lighting;
+
+// Use imported functions
+float3 directLight = Lighting::calculateDirectLighting(...);
+----
+
+Standard HLSL lacks this module system, making it harder to organize large shader codebases.
+
+=== Advanced Metaprogramming
+
+Slang offers powerful compile-time metaprogramming capabilities:
+
+[source,slang]
+----
+// Compile-time reflection
+struct Material {
+    float4 baseColor;
+    float roughness;
+    float metallic;
+};
+// Get all fields of a type at compile time
+__generic<T>
+void bindMaterial(ParameterBlock<T> block, Material material) {
+    __for(field in getFields(T)) {
+        block.setField(field.name, getField(material, field.name));
+    }
+}
+----
+
+=== Resource Binding Model
+
+Slang introduces a more flexible resource binding model:
+
+[source,slang]
+----
+// Parameter block concept
+ParameterBlock<LightingParams> lightingParams;
+// Accessing resources
+Texture2D albedoMap = lightingParams.albedoMap;
+----
+
+This provides better organization and more flexible binding than HLSL's register-based approach.
+
+=== Syntax Differences
+
+While Slang maintains HLSL syntax compatibility, it introduces some new syntax elements:
+
+* *Module declarations*: `module ModuleName;`
+* *Import statements*: `import ModuleName;`
+
+=== Compilation Differences
+
+Slang provides its own compiler (`slangc`) and runtime compilation API with different capabilities than the HLSL compiler:
+
+* *Multi-target compilation*: Compile the same shader for multiple graphics APIs
+* *Cross-compilation*: Generate code for different shader stages from a single source
+* *Built-in reflection*: Generate reflection data during compilation
+* *Shader linking*: Link multiple shader modules together
+* *Diagnostic quality*: More detailed error messages and warnings
+
+Example of multi-target compilation:
+
+[source,bash]
+----
+slangc -profile glsl_spirv -entry main -stage vertex shader.slang -o shader.vert.spv
```

Comment:
I'm not sure if this is valid, highlighting to see if some other reviewer has opinions on this. I've used
`slangc -target spirv shader.slang -o shader.vert.spv` instead. Entry points can be marked in the source with an attribute like `[shader("compute")]` in front of the function. The `-entry main -stage vertex` is still valid though.

Reply:
Exactly. That command line looks weird. Why use `glsl_spirv` when targeting Vulkan? The SPIR-V targets like `spirv_1_4` make more sense imo.